File: game-server/src/main/java/org/forfun/mmorpg/game/util/dirtywords/DirtyWordsReader.java
package org.forfun.mmorpg.game.util.dirtywords;

import lombok.Getter;
import org.apache.commons.lang3.StringUtils;
import org.springframework.core.io.FileSystemResource;
import org.springframework.core.io.Resource;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class DirtyWordsReader {

    private String ENCODING = "UTF-8";
    @Getter
    private final Map<Character, List<String>> dirtyWords = new HashMap<>();
    @Getter
    private int wordCount = 0;

    DirtyWordsReader() throws IOException {
        Set<String> words = readSensitiveWordPool();
        this.initWordsStore(words);
    }

    private Set<String> readSensitiveWordPool() throws IOException {
        Set<String> words = new HashSet<>();
        Resource resource = new FileSystemResource("config/SensitiveWord.txt");
        File file = resource.getFile();
        try (InputStreamReader read = new InputStreamReader(new FileInputStream(file), ENCODING)) {
            if (file.isFile() && file.exists()) {
                BufferedReader bufferedReader = new BufferedReader(read);
                String word;
                while ((word = bufferedReader.readLine()) != null) {
                    if (StringUtils.isEmpty(word)) {
                        continue;
                    }
                    wordCount++;
                    words.add(word.trim());
                }
            } else {
                throw new FileNotFoundException("dirty words file is empty");
            }
        } catch (Exception e) {
            throw e;
        }
        return words;
    }

    private void initWordsStore(Set<String> words) {
        for (String word : words) {
            if (StringUtils.isEmpty(word)) {
                continue;
            }
            char first = word.charAt(0);
            List<String> sameFirst = dirtyWords.computeIfAbsent(first, k -> new ArrayList<>());
            sameFirst.add(word);
        }
    }

}

File: game-server/src/main/java/org/forfun/mmorpg/game/util/dirtywords/DirtyWordsReader.kt
package org.forfun.mmorpg.game.util.dirtywords

import org.apache.commons.lang3.StringUtils
import org.springframework.core.io.FileSystemResource
import org.springframework.core.io.Resource
import java.io.BufferedReader
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.IOException
import java.io.InputStreamReader
import java.util.ArrayList
import java.util.HashMap
import java.util.HashSet

class DirtyWordsReader () {

    private val ENCODING = "UTF-8"
    private val dirtyWords: MutableMap<Char, MutableList<String>> = HashMap()
    private var wordCount: Int = 0

    init {
        val words = readSensitiveWordPool()
        initWordsStore(words)
    }

    @Throws(IOException::class)
    private fun readSensitiveWordPool(): Set<String> {
        val words: MutableSet<String> = HashSet()
        val resource: Resource = FileSystemResource("config/SensitiveWord.txt")
        val file = resource.file
        InputStreamReader(FileInputStream(file), ENCODING).use { read ->
            BufferedReader(read).use { bufferedReader ->
                if (file.isFile && file.exists()) {
                    var word: String? = bufferedReader.readLine()
                    while (word != null) {
                        if (StringUtils.isEmpty(word)) {
                            word = bufferedReader.readLine()
                            continue
                        }
                        wordCount++
                        words.add(word.trim { it <= ' ' })
                        word = bufferedReader.readLine()
                    }
                } else {
                    throw FileNotFoundException("dirty words file is empty")
                }
            }
        }
        return words
    }

    private fun initWordsStore(words: Set<String>) {
        for (word in words) {
            if (StringUtils.isEmpty(word)) {
                continue
            }
            val first = word[0]
            val sameFirst = dirtyWords.computeIfAbsent(first) { ArrayList() }
            sameFirst.add(word)
        }
    }

    fun getDirtyWords(): Map<Char, MutableList<String>> {
        return dirtyWords
    }

    fun getWordCount(): Int {
        return wordCount
    }

    companion object {
        @Throws(IOException::class)
        fun create(): DirtyWordsReader {
            return DirtyWordsReader()
        }
    }
}
